pkg/client/command_line.go

```diff
@@ -15,62 +16,215 @@ func AddArmadaApiConnectionCommandlineArgs(rootCmd *cobra.Command) {
+// contexts exists, and that the current context is defined. Other methods on armadaClientConfig
+// are themselves responsible for performing any further validation.
+func (config *armadaClientConfig) validate() error {
+	if config == nil {
+		return fmt.Errorf("[armadaClientConfig.validate] config is nil")
+	}
+
+	if config.Users != nil {
+		for name, user := range config.Users {
+			if name == "" {
+				return fmt.Errorf("[armadaClientConfig.validate] empty user names not allowed")
+			}
+			if user == nil {
+				return fmt.Errorf("[armadaClientConfig.validate] user %s is nil", name)
+			}
+		}
+	}
+
+	if config.Clusters != nil {
+		for name, cluster := range config.Clusters {
+			if name == "" {
+				return fmt.Errorf("[armadaClientConfig.validate] empty cluster names not allowed")
+			}
+			if cluster == nil {
+				return fmt.Errorf("[armadaClientConfig.validate] cluster %s is nil", name)
+			}
+		}
+	}
+
+	if config.Contexts != nil {
+		if config.Users == nil {
+			return fmt.Errorf("[armadaClientConfig.validate] no users provided")
+		}
+		if config.Clusters == nil {
+			return fmt.Errorf("[armadaClientConfig.validate] no clusters provided")
+		}
+
+		for name, context := range config.Contexts {
+			if name == "" {
+				return fmt.Errorf("[armadaClientConfig.validate] empty context names not allowed")
+			}
+			if context == nil {
+				return fmt.Errorf("[armadaClientConfig.validate] context %s is nil", name)
+			}
+			if _, ok := config.Clusters[context.Cluster]; !ok {
+				return fmt.Errorf("[armadaClientConfig.validate] could not find cluster %s referenced by context %s", context.Cluster, name)
+			}
+			if _, ok := config.Users[context.User]; !ok {
+				return fmt.Errorf("[armadaClientConfig.validate] could not find user %s referenced by context %s", context.User, name)
+			}
+		}
+	}
+
+	if config.CurrentContext != "" {
+		if config.Clusters == nil {
+			return fmt.Errorf("[armadaClientConfig.validate] could not find current context %s", config.CurrentContext)
+		}
+		if _, ok := config.Contexts[config.CurrentContext]; !ok {
+			return fmt.Errorf("[armadaClientConfig.validate] could not find current context %s", config.CurrentContext)
+		}
+	}
+
+	return nil
+}
+
+// ConfigYamlString returns a string representation of the merged configuration stored in viper.
+func (config *armadaClientConfig) YamlString() (string, error) {
+	bs, err := yaml.Marshal(config)
+	if err != nil {
+		return "", fmt.Errorf("[armadaClientConfig.YamlString] unable to marshal config to yaml: %s", err)
+	}
+	return string(bs), nil
+}
+
+// LoadClientConfig reads in config from the file at configPath, or, if configPath is the empty
+// string, from $HOME/.armada/config, and from environment variables, using viper.
+func LoadClientConfig(configPath string) (*armadaClientConfig, error) {
+
+	// if no file is provided (i.e., if configPath is the empty string), default to $HOME/.armada/config
+	if len(configPath) == 0 {
```

Comment:
can we log which files we resolved or does viper do that for us?

Reply:
Doesn't seem like viper does it for us. Instead, I've added code that adds the found config files to a slice.